### PR TITLE
Release 3.4.1

### PR DESCRIPTION
**Pull request by AdamBJohnsonx**

### CODE REVIEW

**Comment by shahzaibj** [comment]
Why is this going inside 3.4.0? that has already been released right

**Comment by AdamBJohnsonx** [comment]
> Why is this going inside 3.4.0? that has already been released right

Yes, I selected the wrong base.  (functionally, they should be identical)

File: common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java [hunk 1]
```diff
@@ -91,6 +96,13 @@
     private static InteractiveTokenCommand sCommand = null;
     private static final CommandResultCache sCommandResultCache = new CommandResultCache();
 
+    private static final TreeSet<String> nonCacheableErrorCodes = new TreeSet(
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java:99]: Any reason we need to use a `TreeSet` vs a `HashSet`? Is there any ordering dependency?

File: common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java [hunk 2]
```diff
@@ -476,7 +488,15 @@ private static boolean eligibleToCache(CommandResult commandResult) {
      * @return
      */
     private static boolean eligibleToCacheException(BaseException exception) {
-        if (exception instanceof IntuneAppProtectionPolicyRequiredException) {
+        final String mErrorCode;
+        if (exception instanceof BrokerCommunicationException) {
+            mErrorCode = ((BrokerCommunicationException) exception).getCategory().toString();
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java:493]: `category` strictly cannot be `null`? This is certain?

Comment (AdamBJohnsonx) [reply common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java:493]: It is dereferenced in the sole constructor of the class.

File: common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java (hunk 1 above)
Comment (AdamBJohnsonx) [reply common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java:99]: For such a small number of items, TreeSet can typically be (slightly) more efficient.

File: common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java (hunk 2 above)
Comment (AdamBJohnsonx) [reply common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java:493]: Actually, we don't need to do this at all... BrokerCommunicationException /is a/ BaseException, and the errorCode is category.toString().

Comment (AdamBJohnsonx) [reply common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java:493]: I'm inclined to leave it alone, though.

**Review by iambmelt — APPROVED** [review]

File: common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java [hunk 3]
```diff
@@ -476,7 +488,15 @@ private static boolean eligibleToCache(CommandResult commandResult) {
      * @return
      */
     private static boolean eligibleToCacheException(BaseException exception) {
-        if (exception instanceof IntuneAppProtectionPolicyRequiredException) {
+        final String mErrorCode;
```
Comment (shahzaibj) [inline common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java:491]:
```suggestion
        final String errorCode;
```

**Review by shahzaibj — APPROVED** [review]

File: common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java [hunk 4]
```diff
@@ -476,7 +488,15 @@ private static boolean eligibleToCache(CommandResult commandResult) {
...
+        if (exception instanceof BrokerCommunicationException) {
+            mErrorCode = ((BrokerCommunicationException) exception).getCategory().toString();
+        } else {
+            mErrorCode = exception.getErrorCode();
```
Comment (AdamBJohnsonx) [inline common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java:495]:
```suggestion
            errorCode = exception.getErrorCode();
```

File: common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java (hunk 2 above)
Comment (AdamBJohnsonx) [inline common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java:493]:
```suggestion
            errorCode = ((BrokerCommunicationException) exception).getCategory().toString();
```

File: common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java [hunk 5]
```diff
@@ -476,7 +488,15 @@ private static boolean eligibleToCache(CommandResult commandResult) {
...
-        if (exception instanceof IntuneAppProtectionPolicyRequiredException) {
+        final String errorCode;
+        if (exception instanceof BrokerCommunicationException) {
+            errorCode = ((BrokerCommunicationException) exception).getCategory().toString();
+        } else {
+            errorCode = exception.getErrorCode();
+        }
+        //TODO : ADO 1373343 Add the whole transient exception category.
+        if (exception instanceof IntuneAppProtectionPolicyRequiredException
+                || nonCacheableErrorCodes.contains(mErrorCode)) {
```
Comment (AdamBJohnsonx) [inline common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java:499]:
```suggestion
                || nonCacheableErrorCodes.contains(errorCode)) {
```

**Comment by AdamBJohnsonx** [comment]
https://identitydivision.visualstudio.com/IDDP/_build/results?buildId=769016&view=results

**Pull request merged**